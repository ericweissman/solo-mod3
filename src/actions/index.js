export const isLoading = (bool) => ({
  type: 'IS_LOADING',
  isLoading: bool
})

export const hasErrored = (error) => ({
  type: 'HAS_ERRORED',
  error
})

export const fetchArtistsSuccess = (artists, genre) => ({
  type: 'FETCH_ARTISTS_SUCCESS',
  artists,
  genre
})

export const getDeepHouseSuccess = (artists) => ({
  type: 'GET_DEEP_HOUSE_SUCCESS',
  artists
})

export const getFutureBassSuccess = (artists) => ({
  type: 'GET_FUTURE_BASS_SUCCESS',
  artists
})

export const getGlitchHopSuccess = (artists) => ({
  type: 'GET_GLITCH_HOP_SUCCESS',
  artists
})

export const getTrapSuccess = (artists) => ({
  type: 'GET_TRAP_SUCCESS',
  artists
})

export const addToFavorites = (artist) => ({
  type: "ADD_TO_FAVORITES",
  artist
})

export const removeFromFavorites = (id) => ({
  type: "REMOVE_FROM_FAVORITES",
  id
})

export const populateFavorites = (artists) => ({
  type: 'POPULATE_FAVORITES',
  artists
})

