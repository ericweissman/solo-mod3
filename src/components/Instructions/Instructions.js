import React from 'react'

const Instructions = () => {
  return (
    <div className='instructions'>
      <h3>Welcome to AARPlur</h3>
      <p>Your one stop shop to learn about all the music that the young whippersnappers are boogying to!</p>
      <h5>Instructions</h5>
      <ul>
        <li>Select one of the music genres using the left-side menu</li>
        <li>If you click on the image on the card, it will play you a popular song from the artist. Make sure you have your volume turned on your computer!</li>
        <li>If you still want to learn more about the artist, click on the More Info button to visit their online encyclopedia page.</li>
      </ul>
      <p>We hope you enjoy learning about this newfangled robot music!</p>
    </div>
  )
}

export default Instructions