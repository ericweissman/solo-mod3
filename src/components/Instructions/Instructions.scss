.instructions {
  margin-left: 250px;
  padding-top: 100px;
  padding-right: 50px;
  display: flex;
  flex-direction: column;
  align-items: center;
  font-family: $secondary-font;
  color: $color-accent-light;
  text-align: center;

  p {
    padding: 15px 0;
  }

  h5 {
    text-decoration: underline;
  }

  li {
    text-align: left;
  }
  h3 {
    font-family: $primary-font;
    font-size: 40px;
    color: $color-accent;
  }
}