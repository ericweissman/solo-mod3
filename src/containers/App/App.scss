.App {
  display: grid;
  grid-template-columns: 185px 1fr;
  background-color: $color-background-main;
  min-height: 850px;
}