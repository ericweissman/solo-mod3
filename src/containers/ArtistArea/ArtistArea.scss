.artist-area {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  padding-left: 185px;
}