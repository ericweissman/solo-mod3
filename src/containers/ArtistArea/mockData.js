export const mockPropsEmpty = {
  glitchHop: [],
  trap: [],
  deepHouse: [],
  futureBass: [],
  isLoading: false,
  favorites: [],
  fetchArtists: jest.fn()
}
export const mockPropsFetchedNoFaves = {
  glitchHop: [{ name: 'test', id: '1234', wiki: 'asfas', favorited: false }],
  trap: [{ name: 'test', id: '22', wiki: 'asfas', favorited: false }],
  deepHouse: [{ name: 'test', id: '33', wiki: 'asfas', favorited: false }],
  futureBass: [{ name: 'test', id: '44', wiki: 'asfas', favorited: false }],
  isLoading: false,
  favorites: [],
  match: { path: '/glitchHop' },
  fetchArtists: jest.fn()
}
export const mockPropsFetchedWithFaves = {
  glitchHop: [{ name: 'test', id: '1234', wiki: 'asfas', favorited: false }],
  trap: [{ name: 'test', id: '22', wiki: 'asfas', favorited: false }],
  deepHouse: [{ name: 'test', id: '33', wiki: 'asfas', favorited: false }],
  futureBass: [{ name: 'test', id: '44', wiki: 'asfas', favorited: false }],
  isLoading: false,
  favorites: ['1234'],
  match: { path: '/glitchHop' },
  fetchArtists: jest.fn()
}
export const glitchHop = {
  glitchHop: [{ name: 'test', id: '1234', wiki: 'asfas', favorited: false }],
  trap: [{ name: 'test', id: '22', wiki: 'asfas', favorited: false }],
  deepHouse: [{ name: 'test', id: '33', wiki: 'asfas', favorited: false }],
  futureBass: [{ name: 'test', id: '44', wiki: 'asfas', favorited: false }],
  isLoading: false,
  favorites: ['1234'],
  match: { path: '/glitchHop' },
  fetchArtists: jest.fn()
}
export const deepHouse = {
  glitchHop: [{ name: 'test', id: '1234', wiki: 'asfas', favorited: false }],
  trap: [{ name: 'test', id: '22', wiki: 'asfas', favorited: false }],
  deepHouse: [{ name: 'test', id: '33', wiki: 'asfas', favorited: false }],
  futureBass: [{ name: 'test', id: '44', wiki: 'asfas', favorited: false }],
  isLoading: false,
  favorites: ['1234'],
  match: { path: '/deepHouse' },
  fetchArtists: jest.fn()
}
export const futureBass = {
  glitchHop: [{ name: 'test', id: '1234', wiki: 'asfas', favorited: false }],
  trap: [{ name: 'test', id: '22', wiki: 'asfas', favorited: false }],
  deepHouse: [{ name: 'test', id: '33', wiki: 'asfas', favorited: false }],
  futureBass: [{ name: 'test', id: '44', wiki: 'asfas', favorited: false }],
  isLoading: false,
  favorites: ['1234'],
  match: { path: '/futureBass' },
  fetchArtists: jest.fn()
}
export const trap = {
  glitchHop: [{ name: 'test', id: '1234', wiki: 'asfas', favorited: false }],
  trap: [{ name: 'test', id: '22', wiki: 'asfas', favorited: false }],
  deepHouse: [{ name: 'test', id: '33', wiki: 'asfas', favorited: false }],
  futureBass: [{ name: 'test', id: '44', wiki: 'asfas', favorited: false }],
  isLoading: false,
  favorites: ['1234'],
  match: { path: '/trap' },
  fetchArtists: jest.fn()
}
export const favorites = {
  glitchHop: [{ name: 'test', id: '1234', wiki: 'asfas', favorited: false }],
  trap: [{ name: 'test', id: '22', wiki: 'asfas', favorited: false }],
  deepHouse: [{ name: 'test', id: '33', wiki: 'asfas', favorited: false }],
  futureBass: [{ name: 'test', id: '44', wiki: 'asfas', favorited: false }],
  isLoading: false,
  favorites: ['1234'],
  match: { path: '/favorites' },
  fetchArtists: jest.fn()
}

export const emptyGH = {
  glitchHop: [],
  trap: [{ name: 'test', id: '22', wiki: 'asfas', favorited: false }],
  deepHouse: [{ name: 'test', id: '33', wiki: 'asfas', favorited: false }],
  futureBass: [{ name: 'test', id: '44', wiki: 'asfas', favorited: false }],
  isLoading: false,
  favorites: ['1234'],
  match: { path: '/glitchHop' },
  fetchArtists: jest.fn()
}
export const emptyTrap = {
  glitchHop: [],
  trap: [],
  deepHouse: [{ name: 'test', id: '33', wiki: 'asfas', favorited: false }],
  futureBass: [{ name: 'test', id: '44', wiki: 'asfas', favorited: false }],
  isLoading: false,
  favorites: ['1234'],
  match: { path: '/trap' },
  fetchArtists: jest.fn()
}
export const emptyDH = {
  glitchHop: [],
  trap: [],
  deepHouse: [{ name: 'test', id: '33', wiki: 'asfas', favorited: false }],
  futureBass: [{ name: 'test', id: '44', wiki: 'asfas', favorited: false }],
  isLoading: false,
  favorites: ['1234'],
  match: { path: '/trap' },
  fetchArtists: jest.fn()
}
export const emptyFB = {
  glitchHop: [],
  trap: [],
  deepHouse: [{ name: 'test', id: '33', wiki: 'asfas', favorited: false }],
  futureBass: [{ name: 'test', id: '44', wiki: 'asfas', favorited: false }],
  isLoading: false,
  favorites: ['1234'],
  match: { path: '/trap' },
  fetchArtists: jest.fn()
}
export const emptyPath = {
  glitchHop: [],
  trap: [],
  deepHouse: [{ name: 'test', id: '33', wiki: 'asfas', favorited: false }],
  futureBass: [{ name: 'test', id: '44', wiki: 'asfas', favorited: false }],
  isLoading: false,
  favorites: ['1234'],
  match: { path: '' },
  fetchArtists: jest.fn()
}