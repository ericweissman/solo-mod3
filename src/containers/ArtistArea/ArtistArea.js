import React, { Component } from 'react'
import * as actions from '../../actions'
import { fetchArtists } from '../../thunks/fetchArtists'
import { connect } from 'react-redux'
import Card from '../../components/Card/Card'
import * as url from '../../utils/urls'
import Loading from '../../components/Loading/Loading'
const key = require('short-id')

class ArtistArea extends Component {

  getArtists = () => {
    switch (this.props.match.path) {
      case '/glitchHop':
        if (this.props.glitchHop.length === 0) {
          this.props.fetchArtists(url.glitchHop, actions.getGlitchHopSuccess)
        }
        return 'glitchHop'
      case '/deepHouse':
        if (this.props.deepHouse.length === 0) {
          this.props.fetchArtists(url.deepHouse, actions.getDeepHouseSuccess)
        }
        return 'deepHouse'
      case '/trap':
        if (this.props.trap.length === 0) {
          this.props.fetchArtists(url.trap, actions.getTrapSuccess)
        }
        return 'trap'
      case '/futureBass':
        if (this.props.futureBass.length === 0) {
          this.props.fetchArtists(url.futureBass, actions.getFutureBassSuccess)
        }
        return 'futureBass'
      case '/favorites':
        return 'favorites'
      default:
        return null
    }
  }

  cardsToDisplay = () => {
    const genre = this.getArtists()
    const cards = this.props[genre].map(artist => {
      return <Card artist={artist} key={key.generate()} />
    })
    const {isLoading} = this.props
    if (isLoading) {
      return <Loading />
    }
    return cards
  }

  render() {
    return (
      <div>
        {this.cardsToDisplay()}
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  glitchHop: state.glitchHop,
  trap: state.trap,
  deepHouse: state.deepHouse,
  futureBass: state.futureBass,
  isLoading: state.isLoading,
  favorites: state.favorites,
})

export const mapDispatchToProps = (dispatch) => ({
  fetchArtists: (url, actionToDispatch) => dispatch(fetchArtists(url, actionToDispatch))
})

export default connect(mapStateToProps, mapDispatchToProps)(ArtistArea)