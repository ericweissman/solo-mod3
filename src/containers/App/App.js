import React, { Component } from 'react';
import { Switch, Route, withRouter } from 'react-router-dom'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import '../../styles/index.scss'
import { populateFavorites } from '../../actions'
import Instructions from '../../components/Instructions/Instructions'
import NotFound from '../../components/NotFound/NotFound'
import ErrorMsg from '../../components/ErrorMsg/ErrorMsg'
import Nav from '../Nav/Nav'
import ArtistArea from '../ArtistArea/ArtistArea'
import Loading from '../../components/Loading/Loading';

export class App extends Component {

  componentDidMount() {
    const favorites = JSON.parse(localStorage.getItem('favorites'))
    if (favorites !== null) {
      this.props.populateFavorites(favorites)
    }
  }

  getRouteForGenre = ({match}) =>{
    return (
      <ArtistArea match={match} />
    )
  }

  render() {
    const { hasErrored } = this.props
    switch (hasErrored) {
      case '':
        return (
          <div className="App">
            <Nav />
            <Switch>
              <Route exact path='/' component={Instructions} />
              <Route path='/glitchHop' render={this.getRouteForGenre} />
              <Route path='/deepHouse' render={this.getRouteForGenre} />
              <Route path='/futureBass' render={this.getRouteForGenre} />
              <Route path='/trap' render={this.getRouteForGenre} />
              <Route path='/favorites' render={this.getRouteForGenre} />
              <Route component={NotFound} />
            </Switch>
          </div>
        );
      default:
        return (
          <ErrorMsg message={hasErrored} />
        )
    }
  }
}

export const mapStateToProps = (state) => ({
  deepHouse: state.deepHouse,
  futureBass: state.futureBass,
  glitchHop: state.glitchHop,
  trap: state.trap,
  hasErrored: state.hasErrored,
})

export const mapDispatchToProps = (dispatch) => ({
  populateFavorites: (artists) => dispatch(populateFavorites(artists))
})


export default withRouter(connect(mapStateToProps, mapDispatchToProps)(App));

App.propTypes = {
  deepHouse: PropTypes.array,
  glitchHop: PropTypes.array,
  trap: PropTypes.array,
  futureBass: PropTypes.array,
  hasErrored: PropTypes.string,
  populateFavorites: PropTypes.func,
}

App.defaultProps = {
  deepHouse: [],
  glitchHop: [],
  trap: [],
  futureBass: [],
  hasErrored: '',
}