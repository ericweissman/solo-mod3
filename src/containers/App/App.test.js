import React from 'react';
import ReactDOM from 'react-dom';
import { App, mapDispatchToProps, mapStateToProps } from './App';
import { populateFavorites } from '../../actions'
import { NotFound } from '../../components/NotFound/NotFound'
import { Instructions} from '../../components/Instructions/Instructions'
import { ArtistArea} from '../ArtistArea/ArtistArea'
import { shallow, mount } from 'enzyme'
import { BrowserRouter, MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux'
import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension'
import { rootReducer } from '../../reducers/index'
import thunk from 'redux-thunk';


const mockProps = {
  deepHouse: [],
  futureBass: [],
  glitchHop: [],
  trap: [],
  hasErrored: false,
  populateFavorites: jest.fn(),
}

const withError = {
  deepHouse: [],
  futureBass: [],
  glitchHop: [],
  trap: [],
  hasErrored: true,
  populateFavorites: jest.fn(),
}

describe('App', () => {
  const store = createStore(rootReducer, composeWithDevTools(applyMiddleware(thunk)))
  let wrapper

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
      , div);
    ReactDOM.unmountComponentAtNode(div);
  });

  describe('snapshots', () => {
    it('should have the correct snapshot when there is no error', () => {
      wrapper = shallow(<App {...mockProps} />)
      expect(wrapper).toMatchSnapshot()
    })
    it('should have the correct snapshot when there is an error', () => {
      wrapper = shallow(<App {...withError} />)
      expect(wrapper).toMatchSnapshot()
    })
  })
  
  describe('Routes', () => {
    it('should render Instructions with the home route', () => {
      const wrapper = mount(
        <Provider store={store}>
          <MemoryRouter initialEntries={['/']}>
            <App />
          </MemoryRouter>
        </Provider>
      )
      expect(wrapper.find(Instructions)).toHaveLength(1)
    })
  })


  describe('mapStateToProps', () => {
    it('should return the correct pieces of state', () => {
      const mockState = {
        deepHouse: [],
        futureBass: [],
        glitchHop: [],
        trap: [],
        hasErrored: false,
        test: 1,
        example: 2
      }

      const expected = {
        deepHouse: [],
        futureBass: [],
        glitchHop: [],
        trap: [],
        hasErrored: false,
      }

      const mappedProps = mapStateToProps(mockState)
      expect(mappedProps).toEqual(expected)
    })

  })

  describe('mapDispatchToProps', () => {
    it('should call dispatch with populateFavorites', () => {
      const mockDispatch = jest.fn()
      const mockFavorites = ['asfasfa', 'gdsgrsdg']
      const actionToDispatch = populateFavorites(mockFavorites)

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.populateFavorites(mockFavorites)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
})
