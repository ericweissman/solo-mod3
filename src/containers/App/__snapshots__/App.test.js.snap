// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App getRouteForGenre should display the glitchHop in Artist area based on path 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`App snapshots should have the correct snapshot when there is an error 1`] = `
<ErrorMsg
  message="Some error"
/>
`;

exports[`App snapshots should have the correct snapshot when there is no error 1`] = `
<div
  className="App"
>
  <Nav />
  <Switch>
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
    <Route
      path="/glitchHop"
      render={[Function]}
    />
    <Route
      path="/deepHouse"
      render={[Function]}
    />
    <Route
      path="/futureBass"
      render={[Function]}
    />
    <Route
      path="/trap"
      render={[Function]}
    />
    <Route
      path="/favorites"
      render={[Function]}
    />
    <Route
      component={[Function]}
    />
  </Switch>
</div>
`;
